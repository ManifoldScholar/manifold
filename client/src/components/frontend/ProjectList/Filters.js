import React, { Component } from "react";
import PropTypes from "prop-types";

export default class ProjectListFilters extends Component {
  static displayName = "ProjectList.Filters";

  static propTypes = {
    updateAction: PropTypes.func,
    subjects: PropTypes.array,
    hideFeatured: PropTypes.bool
  };

  filterChange = event => {
    let filter = {};
    const value = event.target.value;
    if (value) {
      switch (value) {
        case "featured":
          filter = { featured: true };
          break;
        case "notFeatured":
          filter = { featured: false };
          break;
        default:
          filter = { subject: value };
          break;
      }
    }
    this.props.updateAction(filter);
  };

  subjectOptions() {
    if (!this.props.subjects) return null;
    return this.props.subjects.map(subject => {
      return (
        <option key={subject.id} value={subject.id}>
          {subject.attributes.name}
        </option>
      );
    });
  }

  featuredOptions() {
    if (this.props.hideFeatured) return null;
    return <option value="featured">Featured Projects</option>;
  }

  render() {
    return (
      <div className="select-browse">
        <select onChange={this.filterChange}>
          <option value="">Show All</option>
          {this.featuredOptions()}
          {this.subjectOptions()}
        </select>
        <i className="manicon manicon-caret-down" />
      </div>
    );
  }
}
