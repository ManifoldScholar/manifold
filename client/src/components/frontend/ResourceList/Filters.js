import React, { Component, PropTypes } from 'react';
import capitalize from 'lodash/capitalize';
import get from 'lodash/get';

export default class ResourceListFilters extends Component {

  static displayName = "ResourceList.Filters";

  static propTypes = {
    kinds: PropTypes.array,
    tags: PropTypes.array,
    filterChangeHandler: PropTypes.func.isRequired,
    initialFilterState: PropTypes.object
  };

  constructor(props) {
    super(props);
    this.state = this.initialState(props.initialFilterState);
    this.setFilters = this.setFilters.bind(this);
    this.resetFilters = this.resetFilters.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    if (get(prevState, 'filter') !== get(this.state, 'filter')) {
      this.props.filterChangeHandler(this.state.filter);
    }
  }

  setFilters(event, label) {
    event.preventDefault();
    const value = (label === "keyword")
      ? event.target.querySelector(".search-input input").value
      : event.target.value;
    const filter = Object.assign({}, this.state.filter);
    const inputs = Object.assign({}, this.state.inputs);
    if (value && label) {
      switch (value) {
        case "default":
          delete filter[label];
          inputs[label] = "default";
          break;
        default:
          filter[label] = value;
          inputs[label] = value;
          break;
      }
      this.setState({ inputs, filter }, this.updateResults);
    }
  }

  setKeywordInput(event) {
    const value = event.target.value;
    const inputs = Object.assign({}, this.state.inputs);
    const filter = Object.assign({}, this.state.filter);
    if (value === "") {
      inputs.keyword = "";
      delete filter.keyword;
    } else {
      inputs.keyword = value;
    }
    this.setState({ inputs, filter });
  }

  initialState(init) {
    return {
      filter: {
      },
      inputs: init ? init : {
        keyword: "",
        kind: "default",
        tag: "default",
        order: "default"
      }
    };
  }

  resetFilters(event) {
    event.preventDefault();
    this.setState(this.initialState());
  }

  render() {
    return (
      <form
        className="form-list-filter"
        onSubmit={event => this.setFilters(event, "keyword")}
      >
        <div className="search-input">
          <button className="search-button" type="submit">
            <i className="manicon manicon-magnify"></i>
          </button>
          <input
            value={this.state.inputs.keyword}
            type="text"
            onChange={event => this.setKeywordInput(event)}
            placeholder="Search"
          />
        </div>
        <div className="select-group">
          <div className="select">
            <select
              onChange={event => this.setFilters(event, "kind")}
              value={this.state.inputs.kind}
            >
              <option value="default">
                Type:
              </option>
              {this.props.kinds ?
                this.props.kinds.map((kind, index) => {
                  return (
                    <option key={index} value={kind}>{capitalize(kind)}</option>
                  );
                })
              : null}
            </select>
            <i className="manicon manicon-caret-down"></i>
          </div>
          <div className="select">
            <select
              onChange={event => this.setFilters(event, "tag")}
              value={this.state.inputs.tag}
            >
              <option value="default">
                Tag:
              </option>
              {this.props.tags ?
                this.props.tags.map((tag, index) => {
                  return (
                    <option key={index} value={tag}>{capitalize(tag)}</option>
                  );
                })
              : null}
            </select>
            <i className="manicon manicon-caret-down"></i>
          </div>
          <div className="select">
            <select
              onChange={event => this.setFilters(event, "order")}
              value={this.state.inputs.order}
            >
              <option value="default">
                Order By:
              </option>
              <option value="title ASC">A-Z</option>
              <option value="title DESC">Z-A</option>
            </select>
            <i className="manicon manicon-caret-down"></i>
          </div>
        </div>
        <button className="reset-button" onClick={this.resetFilters}>
          {'Reset Search + Filters'}
        </button>
      </form>
    );
  }
}
