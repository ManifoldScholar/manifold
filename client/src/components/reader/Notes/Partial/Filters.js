import React, { Component } from "react";
import PropTypes from "prop-types";

export default class Filters extends Component {
  static displayName = "Notes.List.Filters";

  static propTypes = {
    filterChangeHandler: PropTypes.func,
    filter: PropTypes.object
  };

  filteredBy(filter, format) {
    return filter.indexOf(format) > -1;
  }

  toggleFormat = (event, format) => {
    event.preventDefault();
    const formatsFilter = this.props.filter.formats;
    if (this.filteredBy(formatsFilter, format)) {
      formatsFilter.splice(formatsFilter.indexOf(format), 1);
    } else {
      formatsFilter.push(format);
    }
    this.props.filterChangeHandler("formats", formatsFilter);
  };

  renderCheckBox(label, format) {
    const formats = this.props.filter.formats;
    const checked = this.filteredBy(formats, format);

    return (
      <label className="checkbox">
        <input
          type="checkbox"
          checked={checked}
          onChange={e => this.toggleFormat(e, format)}
        />
        <div className="control-indicator">
          <i className="manicon manicon-check" />
        </div>
        {label}
      </label>
    );
  }

  render() {
    return (
      <nav className="filters">
        <label className="label">Show your:</label>
        <div className="checkbox-group">
          {this.renderCheckBox("Highlights", "highlight")}
          {this.renderCheckBox("Annotations", "annotation")}
        </div>
      </nav>
    );
  }
}
