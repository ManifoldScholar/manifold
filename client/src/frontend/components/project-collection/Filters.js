import React, { Component } from "react";
import PropTypes from "prop-types";
import omitBy from "lodash/omitBy";
import uniqueId from "lodash/uniqueId";
import Utility from "global/components/utility";

export default class ProjectCollectionFilters extends Component {
  static defaultProps = {
    searchId: uniqueId("filters-search-")
  };

  static displayName = "ProjectCollection.Filters";

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.initialFilterState !== prevState.filters) {
      return Object.assign({}, nextProps.initialFilterState);
    }

    return null;
  }

  static propTypes = {
    filterChangeHandler: PropTypes.func.isRequired,
    initialFilterState: PropTypes.object,
    searchId: PropTypes.string
  };

  constructor(props) {
    super(props);
    this.state = this.initialState(props.initialFilterState);
  }

  setFilters = (event, label) => {
    event.preventDefault();
    const value = event.target.value;
    const filters = Object.assign({}, this.state.filters);
    filters[label] = value;
    if (label === "keyword") return this.setState({ filters });
    this.setState({ filters }, this.updateResults);
  };

  initialState(init) {
    const filters = Object.assign({}, init);
    return { filters };
  }

  resetFilters = event => {
    event.preventDefault();
    this.setState(this.initialState(), this.updateResults);
  };

  updateResults = event => {
    if (event) event.preventDefault();
    const filter = omitBy(this.state.filters, value => value === "");
    this.props.filterChangeHandler(filter);
  };

  render() {
    return (
      <form
        className="entity-section-wrapper__tools form-list-filter"
        onSubmit={this.updateResults}
      >
        <div className="search-input">
          <button className="search-button" type="submit">
            <span className="screen-reader-text">Search…</span>
            <Utility.IconComposer
              iconClass="search-icon"
              icon="search16"
              size={20}
            />
          </button>
          <label htmlFor={this.props.searchId} className="screen-reader-text">
            Enter Search Criteria
          </label>
          <input
            value={this.state.filters.keyword || ""}
            type="text"
            id={this.props.searchId}
            onChange={event => this.setFilters(event, "keyword")}
            placeholder="Search…"
          />
        </div>
        <div className="select-group inline">
          <div className="select">
            <select
              onChange={event => this.setFilters(event, "order")}
              value={this.state.filters.order || ""}
            >
              <option value="">Sort</option>
              <option value="sort_title ASC">A-Z</option>
              <option value="sort_title DESC">Z-A</option>
            </select>
            <Utility.IconComposer
              icon="disclosureDown16"
              size={20}
              iconClass="select__icon"
            />
          </div>
        </div>
        <button className="reset-button" onClick={this.resetFilters}>
          {"Reset Search + Filters"}
        </button>
      </form>
    );
  }
}
