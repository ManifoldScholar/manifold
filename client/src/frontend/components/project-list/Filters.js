import React, { Component } from "react";
import PropTypes from "prop-types";
import omitBy from "lodash/omitBy";
import uniqueId from "lodash/uniqueId";
import Utility from "global/components/utility";

export default class ProjectListFilters extends Component {
  static defaultProps = {
    searchId: uniqueId("filters-search-")
  };

  static displayName = "ProjectList.Filters";

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.initialFilterState !== prevState.filters) {
      return Object.assign({}, nextProps.initialFilterState);
    }

    return null;
  }

  static propTypes = {
    filterChangeHandler: PropTypes.func,
    initialFilterState: PropTypes.object,
    subjects: PropTypes.array,
    hideFeatured: PropTypes.bool,
    searchId: PropTypes.string
  };

  constructor(props) {
    super(props);
    this.state = this.initialState(props.initialFilterState);
  }

  /* eslint-disable react/no-access-state-in-setstate */
  setFilters = (event, label) => {
    event.preventDefault();
    const value = event.target.value;
    let filters = Object.assign({}, this.state.filters);
    filters[label] = value;
    if (label === "keyword") return this.setState({ filters });
    if (label === "order") {
      return this.setState({ filters }, this.updateResults);
    }
    switch (value) {
      case "featured":
        filters = { featured: true };
        break;
      default:
        filters = { subject: value };
        break;
    }
    this.setState({ filters }, this.updateResults);
  };

  resetFilters = event => {
    event.preventDefault();
    this.setState(this.initialState(), this.updateResults);
  };

  updateResults = event => {
    if (event) event.preventDefault();
    const filter = omitBy(this.state.filters, value => value === "");
    this.props.filterChangeHandler(filter);
  };

  featuredOptions() {
    if (this.props.hideFeatured) return null;
    return <option value="featured">Featured Projects</option>;
  }

  filterValue() {
    const { featured, subject } = this.state.filters;
    if (featured === true || featured === "true") return "featured";
    if (subject) return subject;
    return "";
  }

  initialState(init) {
    const filters = Object.assign({}, init);
    return { filters };
  }

  subjectOptions() {
    if (!this.props.subjects || this.props.subjects.length === 0) return null;
    return this.props.subjects.map(subject => {
      return (
        <option key={subject.id} value={subject.id}>
          {subject.attributes.name}
        </option>
      );
    });
  }

  renderFilters() {
    if (!this.featuredOptions() && !this.subjectOptions()) return null;

    return (
      <div className="select">
        <select value={this.filterValue()} onChange={this.setFilters}>
          <option value="">Show All</option>
          {this.featuredOptions()}
          {this.subjectOptions()}
        </select>
        <Utility.IconComposer
          icon="disclosureDown16"
          size={20}
          iconClass="select__icon"
        />
      </div>
    );
  }

  renderSearch() {
    return (
      <div className="search-input">
        <button className="search-button" type="submit">
          <span className="screen-reader-text">Search…</span>
          <Utility.IconComposer
            className="search-icon"
            icon="search16"
            size={20}
          />
        </button>
        <label htmlFor={this.props.searchId} className="screen-reader-text">
          Enter Search Criteria
        </label>
        <input
          value={this.state.filters.keyword || ""}
          type="text"
          id={this.props.searchId}
          onChange={event => this.setFilters(event, "keyword")}
          placeholder="Search…"
        />
      </div>
    );
  }

  renderSort() {
    return (
      <div className="select">
        <select
          onChange={event => this.setFilters(event, "order")}
          value={this.state.filters.order || ""}
        >
          <option value="">Sort</option>
          <option value="sort_title ASC">A-Z</option>
          <option value="sort_title DESC">Z-A</option>
        </select>
        <Utility.IconComposer
          icon="disclosureDown16"
          size={20}
          iconClass="select__icon"
        />
      </div>
    );
  }

  render() {
    return (
      <form
        className="form-list-filter entity-section-wrapper__tools"
        onSubmit={this.updateResults}
      >
        {this.renderSearch()}
        <div className="select-group inline">
          {this.renderSort()}
          {this.renderFilters()}
        </div>
        <button
          className="reset-button"
          onClick={this.resetFilters}
          type="button"
        >
          {"Reset Search + Filters"}
        </button>
      </form>
    );
  }
}
