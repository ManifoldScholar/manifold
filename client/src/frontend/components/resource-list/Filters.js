import React, { Component } from "react";
import PropTypes from "prop-types";
import capitalize from "lodash/capitalize";
import omitBy from "lodash/omitBy";
import labelId from "helpers/labelId";
import Utility from "global/components/utility";

export default class ResourceListFilters extends Component {
  static defaultProps = {
    searchId: labelId("filters-search-")
  };

  static displayName = "ResourceList.Filters";

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.initialFilterState !== prevState.filters) {
      return Object.assign({}, nextProps.initialFilterState);
    }

    return null;
  }

  static propTypes = {
    kinds: PropTypes.array,
    tags: PropTypes.array,
    filterChangeHandler: PropTypes.func.isRequired,
    initialFilterState: PropTypes.object,
    searchId: PropTypes.string
  };

  constructor(props) {
    super(props);
    this.state = this.initialState(props.initialFilterState);
  }

  setFilters = (event, label) => {
    event.preventDefault();
    const value = event.target.value;
    const filters = Object.assign({}, this.state.filters);
    filters[label] = value;
    if (label === "keyword") return this.setState({ filters });
    this.setState({ filters }, this.updateResults);
  };

  initialState(init) {
    const filters = Object.assign({}, init);
    return { filters };
  }

  resetFilters = event => {
    event.preventDefault();
    this.setState(this.initialState(), this.updateResults);
  };

  updateResults = event => {
    if (event) event.preventDefault();
    const filter = omitBy(this.state.filters, value => value === "");
    this.props.filterChangeHandler(filter);
  };

  render() {
    return (
      <form className="form-list-filter" onSubmit={this.updateResults}>
        <div className="search-input">
          <button className="search-button" type="submit">
            <span className="screen-reader-text">Search</span>
            <Utility.IconComposer
              className="search-icon"
              icon="search16"
              size={20}
            />
          </button>
          <label htmlFor={this.props.searchId} className="screen-reader-text">
            Enter Search Criteria
          </label>
          <input
            value={this.state.filters.keyword || ""}
            type="text"
            id={this.props.searchId}
            onChange={event => this.setFilters(event, "keyword")}
            placeholder="Search"
          />
        </div>
        <div className="select-group inline">
          <div className="select">
            <select
              onChange={event => this.setFilters(event, "kind")}
              value={this.state.filters.kind || ""}
            >
              <option value="">Type:</option>
              {this.props.kinds
                ? this.props.kinds.map(kind => {
                    return (
                      <option key={kind} value={kind}>
                        {capitalize(kind)}
                      </option>
                    );
                  })
                : null}
            </select>
            <i className="manicon manicon-caret-down" aria-hidden="true" />
          </div>
          <div className="select">
            <select
              onChange={event => this.setFilters(event, "tag")}
              value={this.state.filters.tag || ""}
            >
              <option value="">Tag:</option>
              {this.props.tags
                ? this.props.tags.map(tag => {
                    return (
                      <option key={tag} value={tag}>
                        {capitalize(tag)}
                      </option>
                    );
                  })
                : null}
            </select>
            <i className="manicon manicon-caret-down" aria-hidden="true" />
          </div>
          <div className="select">
            <select
              onChange={event => this.setFilters(event, "order")}
              value={this.state.filters.order || ""}
            >
              <option value="default">Sort By:</option>
              <option value="sort_title ASC">A-Z</option>
              <option value="sort_title DESC">Z-A</option>
            </select>
            <i className="manicon manicon-caret-down" aria-hidden="true" />
          </div>
        </div>
        <button className="reset-button" onClick={this.resetFilters}>
          {"Reset Search + Filters"}
        </button>
      </form>
    );
  }
}
