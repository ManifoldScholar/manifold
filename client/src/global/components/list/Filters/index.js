export { default as ListFilters } from "./Group";
