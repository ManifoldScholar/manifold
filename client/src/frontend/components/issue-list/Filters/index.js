import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { ListFilters } from "global/components/list";
import useFilterTypes from "./hooks/useFilterTypes";

export default function IssueListFilters(props) {
  const {
    filterChangeHandler,
    initialFilterState,
    resetFilterState,
    subjects,
    hideFeatured = false
  } = props;

  const [filters, setFilters] = useState(initialFilterState);
  const [search, setSearch] = useState("");
  // TODO: update this when using filters other than search
  // const showResetButton = !!Object.values(filters).filter(Boolean).length;
  const showResetButton = !!filters.keyword;

  useEffect(() => {
    filterChangeHandler(filters);
  }, [filters, filterChangeHandler]);

  const updateFilters = (e, label) => {
    e.preventDefault();

    switch (label) {
      case "keyword":
        return setFilters({ ...filters, [label]: search });
      case "order":
        return setFilters({ ...filters, [label]: e.target.value });
      default:
        if (e.target.value === "featured") {
          return setFilters({ featured: true });
        }
        return setFilters({ subject: e.target.value });
    }
  };

  const filterTypes = useFilterTypes(
    filters,
    setFilters,
    updateFilters,
    subjects,
    hideFeatured,
    { sort: false, featuredAndSubject: false }
  );

  const resetFilters = () => {
    const newState = resetFilterState
      ? { ...resetFilterState }
      : initialFilterState;
    setFilters(newState);
  };

  const searchProps = {
    value: search,
    onChange: e => setSearch(e.target.value)
  };

  return (
    <ListFilters
      searchProps={searchProps}
      filters={filterTypes}
      onSubmit={e => updateFilters(e, "keyword")}
      onReset={resetFilters}
      showResetButton={showResetButton}
    />
  );
}

IssueListFilters.displayName = "IssueList.Filters";

IssueListFilters.propTypes = {
  filterChangeHandler: PropTypes.func,
  initialFilterState: PropTypes.object,
  resetFilterState: PropTypes.object,
  subjects: PropTypes.array,
  hideFeatured: PropTypes.bool
};
