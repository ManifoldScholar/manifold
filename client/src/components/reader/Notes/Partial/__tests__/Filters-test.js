import React from "react";
import renderer from "react-test-renderer";
import Filters from "../Filters";

describe("Reader.Notes.Partial.Group Component", () => {
  const props = {
    filter: {
      formats: ["highlight", "annotation", "bookmark"]
    },
    filterChangeHandler: jest.fn()
  };

  const component = renderer.create(<Filters {...props} />);

  it("renders correctly", () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("doesn't render to null", () => {
    let tree = component.toJSON();
    expect(tree).not.toBe(null);
  });
});
