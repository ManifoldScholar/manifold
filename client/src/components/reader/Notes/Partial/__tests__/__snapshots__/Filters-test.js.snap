// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reader.Notes.Partial.Group Component renders correctly 1`] = `
<nav
  className="filters"
>
  <h4
    className="label"
  >
    Show your:
  </h4>
  <div
    className="checkbox-group"
  >
    <label
      className="checkbox"
      htmlFor="highlight-checkbox"
    >
      <input
        checked={true}
        id="highlight-checkbox"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        aria-hidden="true"
        className="control-indicator"
      >
        <i
          className="manicon manicon-check"
        />
      </div>
      Highlights
    </label>
    <label
      className="checkbox"
      htmlFor="annotation-checkbox"
    >
      <input
        checked={true}
        id="annotation-checkbox"
        onChange={[Function]}
        type="checkbox"
      />
      <div
        aria-hidden="true"
        className="control-indicator"
      >
        <i
          className="manicon manicon-check"
        />
      </div>
      Annotations
    </label>
  </div>
</nav>
`;
